example/src/main/java/io/netty/example/dns/doh/DoHClient.java

```diff
@@ -0,0 +1,77 @@
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.example.dns.doh;
+
+import io.netty.bootstrap.Bootstrap;
+import io.netty.channel.Channel;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.nio.NioSocketChannel;
+import io.netty.handler.codec.dns.DefaultDnsQuery;
+import io.netty.handler.codec.dns.DefaultDnsQuestion;
+import io.netty.handler.codec.dns.DnsOpCode;
+import io.netty.handler.codec.dns.DnsQuery;
+import io.netty.handler.codec.dns.DnsRecordType;
+import io.netty.handler.codec.dns.DnsSection;
+import io.netty.handler.ssl.ApplicationProtocolConfig;
+import io.netty.handler.ssl.ApplicationProtocolNames;
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+
+import java.net.URL;
+import java.util.concurrent.TimeUnit;
+
+public final class DoHClient {
+
+    private static final String QUERY_DOMAIN = "netty.io";
+
+    private DoHClient() {
+    }
+
+    public static void main(String[] args) throws Exception {
+        URL url = new URL("https://dns.google/dns-query");
+        EventLoopGroup group = new NioEventLoopGroup();
+
+        try {
+            final SslContext sslContext = SslContextBuilder.forClient()
+                    .applicationProtocolConfig(new ApplicationProtocolConfig(
+                            ApplicationProtocolConfig.Protocol.ALPN,
+                            ApplicationProtocolConfig.SelectorFailureBehavior.NO_ADVERTISE,
+                            ApplicationProtocolConfig.SelectedListenerFailureBehavior.ACCEPT,
+                            ApplicationProtocolNames.HTTP_2,
+                            ApplicationProtocolNames.HTTP_1_1))
+                    .build();
+
+            Bootstrap b = new Bootstrap();
+            b.group(group)
+                    .channel(NioSocketChannel.class)
+                    .handler(new Initializer(sslContext, url));
+
+            Channel ch = b.connect(url.getHost(), url.getDefaultPort()).sync().channel();
+
+            // Wait for application protocol negotiation to finish
+            ch.pipeline().get(ALPNHandler.class).promise().sync();
+
+            // RFC 8484 recommends ID 0 [https://tools.ietf.org/html/rfc8484#section-4.1]
+            DnsQuery query = new DefaultDnsQuery(0, DnsOpCode.QUERY);
+            query.setRecord(DnsSection.QUESTION, new DefaultDnsQuestion(QUERY_DOMAIN, DnsRecordType.A));
```

Comment:
What's the plan for querying A and AAAA records here? Is it enough that a client can take this library and make a A query and a AAAA query a take the first result?

Reply:
Client can pass whatever Record type he wants. All we do is take the DNS data from DnsQuery and full it into HTTP request (in URL in case of GET request and in the body in case of POST).

Lemme add few more queries as well for better insights.

Reply:
Basically I'm curious how this fits into a Netty based server pipeline that needs to do DNS lookups?  This example makes it clear what it looks like to write a simple command line that does IPv4 lookups.